src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php

```diff
@@ -74,30 +65,33 @@ public function __construct(string $coverageDir)
      * @throws NoLineExecuted
      * @throws CoverageDoesNotExistException
      *
-     * @return CoverageFileData[]
+     * @return iterable<CoveredFileData>
      */
-    public function parse(string $coverageXmlContent): array
+    public function parseLazy(string $coverageXmlContent): iterable
     {
         $xPath = self::createXPath($coverageXmlContent);
 
         self::assertHasCoverage($xPath);
 
-        $nodes = $xPath->query('//file');
-
-        $projectSource = self::getProjectSource($xPath);
-
-        $data = [];
-
-        foreach ($nodes as $node) {
-            $relativeFilePath = $node->getAttribute('href');
+        return $this->parseNodes($xPath);
+    }
 
-            $this->processCoverageFile($relativeFilePath, $projectSource, $data);
-        }
+    /**
+     * @deprecated in favor of parseLazy
+     *
+     * @return array<string, CoverageFileData>
+     */
+    public function parse(string $coverageXmlContent): array
+    {
+        $coverage = take($this->parseLazy($coverageXmlContent))
+            ->map(static function (CoveredFileData $data) {
+                yield $data->getSplFileInfo()->getRealPath() => $data->getCoverageFileData();
+            });
 
-        return $data;
+        return iterator_to_array($coverage, true);
     }
 
-    private static function createXPath(string $coverageContent): SafeDOMXPath
+    public static function createXPath(string $coverageContent): SafeDOMXPath
```

Comment:
Maybe this piece and its `removeNamespace()` should be moved in a `XPathFactory` utility class? (in the same namespace)